File: java/hashtable/src/main/java/SimpleHashTable.java
import java.util.LinkedList;

public class SimpleHashTable<K, V> implements HashTable<K, V> {
    
    // should be much bigger and prime
    // doing small for testing
    private static final int TABLE_SIZE = 10;

    private final LinkedList<Entry<K, V>>[] table;

    @SuppressWarnings("unchecked")
    public SimpleHashTable() {
        table = new LinkedList[TABLE_SIZE];
        for (int i = 0; i < TABLE_SIZE; i++) {
            table[i] = new LinkedList<>();
        }
    }

    private static class Entry<K, V> {
        K key;
        V value;

        Entry(K key, V value) {
            this.key = key;
            this.value = value;
        }
    }

    private int hash(K key) {
        return Math.abs(key.hashCode() % TABLE_SIZE);
    }

    @Override
    public void put(K key, V value) {
        int index = hash(key);
        LinkedList<Entry<K, V>> bucket = table[index];

        for (Entry<K, V> entry : bucket) {
            if (entry.key.equals(key)) {
                entry.value = value;
                return;
            }
        }
        bucket.add(new Entry<>(key, value));
    }

    @Override
    public V get(K key) {
        int index = hash(key);
        LinkedList<Entry<K, V>> bucket = table[index];

        for (Entry<K, V> entry : bucket) {
            if (entry.key.equals(key)) {
                return entry.value;
            }
        }
        return null;
    }

    @Override
    public boolean remove(K key) {
        int index = hash(key);
        LinkedList<Entry<K, V>> bucket = table[index];

        for (Entry<K, V> entry : bucket) {
            if (entry.key.equals(key)) {
                return bucket.remove(entry);
            }
        }
        return false;
    }

    @Override
    public boolean isEmpty() {
        for (LinkedList<Entry<K, V>> bucket : table) {
            if (!bucket.isEmpty()) {
                return false;
            }
        }
        return true;
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < TABLE_SIZE; i++) {
            sb.append(String.format("Bucket[%d]: ", i));
            for (Entry<K, V> entry : table[i]) {
                sb.append(String.format("(%s:%s) ", entry.key, entry.value));
            }
            sb.append("\n");
        }
        return sb.toString();
    }
}

File: kotlin/hashtable/src/main/kotlin/SimpleHashTable.kt
// size should prime and bigger
// using smaller for testing
class SimpleHashTable<K, V>(private val size: Int = 10) : HashTable<K, V> {
    
    private val table: Array<MutableList<Entry<K, V>>> = Array(size) { mutableListOf() }

    // very bad hashing but for testing is fine.
    // should use better hashing algorithm
    // good hashing algorithm should distribute keys uniformly
    private fun hash(key: K): Int = 
        Math.abs(key.hashCode() % size)

    override fun put(key: K, value: V) {
        val index = hash(key)
        val bucket = table[index]
        
        val existingEntry = bucket.find { it.key == key }
        if (existingEntry != null) {
            existingEntry.value = value
        } else {
            bucket.add(Entry(key, value))
        }
    }

    override fun get(key: K): V? {
        val index = hash(key)
        return table[index]
            .find { it.key == key }
            ?.value
    }

    override fun remove(key: K): Boolean {
        val index = hash(key)
        val bucket = table[index]
        return bucket.removeIf { it.key == key }
    }

    override fun isEmpty(): Boolean =
        table.all { it.isEmpty() }

    override fun toString(): String =
        table.mapIndexed { index, bucket ->
            val entries = bucket.joinToString(" ") { 
                "(${it.key}:${it.value})" 
            }
            "Bucket[$index]: $entries"
        }.joinToString("\n")
}